Framework/Algorithms/src/AbsorptionCorrectionPaalmanPings.cpp

```diff
@@ -0,0 +1,486 @@
+#include "MantidHistogramData/Interpolate.h"
+#include "MantidKernel/BoundedValidator.h"
+#include "MantidKernel/CompositeValidator.h"
+#include "MantidKernel/Unit.h"
+#include "MantidKernel/UnitFactory.h"
+
+namespace Mantid {
+namespace Algorithms {
+
+// Register the algorithm into the AlgorithmFactory
+DECLARE_ALGORITHM(AbsorptionCorrectionPaalmanPings)
+
+using namespace API;
+using namespace Geometry;
+using HistogramData::interpolateLinearInplace;
+using namespace Kernel;
+using namespace Mantid::DataObjects;
+
+namespace {
+// the maximum number of elements to combine at once in the pairwise summation
+constexpr size_t MAX_INTEGRATION_LENGTH{1000};
+
+inline size_t findMiddle(const size_t start, const size_t stop) {
+  auto half =
+      static_cast<size_t>(floor(.5 * (static_cast<double>(stop - start))));
+  return start + half;
+}
+
+} // namespace
+
+AbsorptionCorrectionPaalmanPings::AbsorptionCorrectionPaalmanPings()
+    : API::Algorithm(), m_inputWS(), m_sampleObject(nullptr),
+      m_containerObject(nullptr), m_L1s(), m_elementVolumes(),
+      m_elementPositions(), m_numVolumeElements(0), m_sampleVolume(0.0),
+      m_containerL1s(), m_containerElementVolumes(),
+      m_containerElementPositions(), m_containerNumVolumeElements(0),
+      m_containerVolume(0.0), m_linearCoefTotScatt(0),
+      m_containerLinearCoefTotScatt(0), m_num_lambda(0), m_xStep(0),
+      m_cubeSide(0.0) {}
+
+void AbsorptionCorrectionPaalmanPings::init() {
+
+  // The input workspace must have an instrument and units of wavelength
+  auto wsValidator = std::make_shared<CompositeValidator>();
+  wsValidator->add<WorkspaceUnitValidator>("Wavelength");
+  wsValidator->add<InstrumentValidator>();
+
+  declareProperty(
+      std::make_unique<WorkspaceProperty<>>("InputWorkspace", "",
+                                            Direction::Input, wsValidator),
+      "The X values for the input workspace must be in units of wavelength");
+  declareProperty(std::make_unique<WorkspaceProperty<WorkspaceGroup>>(
+                      "OutputWorkspace", "", Direction::Output),
+                  "Output workspace name");
+
+  auto positiveInt = std::make_shared<BoundedValidator<int64_t>>();
+  positiveInt->setLower(1);
+  declareProperty(
+      "NumberOfWavelengthPoints", static_cast<int64_t>(EMPTY_INT()),
+      positiveInt,
+      "The number of wavelength points for which the numerical integral is\n"
+      "calculated (default: all points)");
+
+  auto moreThanZero = std::make_shared<BoundedValidator<double>>();
+  moreThanZero->setLower(0.001);
+  declareProperty("ElementSize", 1.0, moreThanZero,
+                  "The size of one side of an integration element cube in mm");
+}
+
+std::map<std::string, std::string>
+AbsorptionCorrectionPaalmanPings::validateInputs() {
+  std::map<std::string, std::string> result;
+
+  // verify that the container information is there if requested
+  API::MatrixWorkspace_const_sptr wksp = getProperty("InputWorkspace");
```

Comment:
This is a good  check to have. Maybe a good idea to have a workspace validator if we expect this to be more common.